sbi/neural_nets/estimators/base.py

```diff
@@ -262,3 +263,144 @@ def forward(self, input: Tensor, condition: Tensor, **kwargs) -> Tensor:
+                `(batch_dim, *input_event_shape)`.
+
+        Returns:
+            Loss of shape (batch_dim,)
+        """
+        pass
+
+    def _check_input_shape(self, input: Tensor):
+        r"""This method checks whether the input has the correct shape.
+
+        Args:
+            input: Inputs to evaluate the log probability on of shape
+                    `(sample_dim_input, batch_dim_input, *event_shape_input)`.
+
+        Raises:
+            ValueError: If the input has a dimensionality that does not match
+                        the expected input dimensionality.
+            ValueError: If the shape of the input does not match the expected
+                        input dimensionality.
+        """
+        input_shape = input.shape
+        exp_input_shape = self.input_shape
+        if len(input_shape) < len(exp_input_shape):
+            raise ValueError(
+                "Dimensionality of input is too small and does not match the "
+                f"expected dimensionality {len(exp_input_shape)}. It should "
+                f"be compatible with the provided input_shape {exp_input_shape}."
+            )
+        else:
+            input_shape = input.shape[-len(self.input_shape) :]
+            if input_shape != exp_input_shape:
+                raise ValueError(
+                    f"Shape of input {input_shape} does not match the "
+                    f"expected input dimensionality {exp_input_shape}, as "
+                    "provided by input_shape. Please reshape it accordingly."
+                )
+
+
+class UnconditionalDensityEstimator(UnconditionalEstimator):
+    r"""Base class for unconditional density estimators.
+
+    The density estimator class is a wrapper around neural networks that allows to
+    evaluate the `log_prob`, `sample`, and provide the `loss` on $x$ values.
+
+    Note:
+        We assume that the input to the density estimator is a tensor of shape
+        (sample_dim, batch_dim, *input_shape), where input_shape is the dimensionality
+        of the input.
+
+    """
+
+    def __init__(self, net: nn.Module, input_shape: torch.Size) -> None:
+        r"""Base class for density estimators.
+
+        Args:
+            net: Neural network or any parameterized model that is used to estimate the
+                probability density of the input.
+            input_shape: Event shape of the input at which the density is being
+                evaluated (and which is also the event_shape of samples).
+        """
+        super().__init__(input_shape)
+        self.net = net
+
+    def log_prob(self, x: Tensor) -> Tensor:
+        r"""Return the log probabilities of the inputs
+
+        Args:
+            x: Inputs to evaluate the log probability on of shape
+                `(sample_dim_input, batch_dim_input, *event_shape_input)`.
+
+        Returns:
+            Sample-wise log probabilities.
+        """
+
+        return self._neural_net.log_prob(x).detach()
```

Comment:
do we really need the detach? better move to eval mode somewhere upstream? How is it down in the `ConditionalDE` class?

Reply:
In fact for conditional density estimators the eval model is activated only at `train`. It is in the `DirectPosterior` class that we see an eval model for the `log_prob` as in [here](https://github.com/sbi-dev/sbi/blob/a1d75559b1e50b3d19c0eac1cabdfa6a23c16ea4/sbi/inference/posteriors/direct_posterior.py#L252). Since we don't have the equivalent of a direct posterior class, I will just put the eval mode in the `log_prob` method for the `UnconditionalDensityEstimator`. Let me know your thoughts.

Reply:
Yes, makes sense 👍